File: src/lib/api/identity.ts
import { invoke } from '@tauri-apps/api/core';
import type { CurrentRepoIdentity, GitIdentity } from '../types';
import { isTauri } from './client';

export async function getCurrentRepoIdentity(path: string): Promise<CurrentRepoIdentity> {
  if (isTauri) {
    return await invoke<CurrentRepoIdentity>('get_current_repo_identity', { path });
  }
  return {
    name: 'Dev User',
    email: 'dev@example.com',
    is_local: false,
  };
}

export async function setRepoIdentity(
  path: string,
  name: string,
  email: string,
  isGlobal: boolean = false
): Promise<void> {
  if (isTauri) {
    await invoke('set_repo_identity', { path, name, email, isGlobal });
  }
}

export async function listIdentityProfiles(): Promise<GitIdentity[]> {
  if (isTauri) {
    return await invoke<GitIdentity[]>('list_identity_profiles');
  }
  return [
    { id: 'default-work', label: 'Công việc (Work)', name: 'Lưu Hồ', email: 'luuho@company.com' },
    { id: 'default-personal', label: 'Cá nhân (Personal)', name: 'Luu Ho', email: 'luuho@gmail.com' },
  ];
}

export async function saveIdentityProfile(identity: GitIdentity): Promise<void> {
  if (isTauri) {
    await invoke('save_identity_profile', { identity });
  }
}

export async function deleteIdentityProfile(id: string): Promise<void> {
  if (isTauri) {
    await invoke('delete_identity_profile', { id });
  }
}

File: src-tauri/src/git/identity.rs
use git2::Repository;
use serde::{Deserialize, Serialize};
use crate::error::{AppError, AppResult};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GitIdentity {
    pub id: String,
    pub label: String, // e.g. "Work (Công ty)", "Personal (Cá nhân)"
    pub name: String,
    pub email: String,
    pub signing_key: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CurrentRepoIdentity {
    pub name: Option<String>,
    pub email: Option<String>,
    pub is_local: bool,
}

/// Read the current active Git author identity for this repository
pub fn get_current_repo_identity(repo: &Repository) -> AppResult<CurrentRepoIdentity> {
    let config = repo.config()?;

    // Check if configured locally in .git/config
    let local_name = config.get_string("user.name").ok();
    let local_email = config.get_string("user.email").ok();

    // Check if it's explicitly in the local config file
    let is_local = if let Ok(entry) = config.get_entry("user.name") {
        entry.level() == git2::ConfigLevel::Local
    } else {
        false
    };

    Ok(CurrentRepoIdentity {
        name: local_name,
        email: local_email,
        is_local,
    })
}

/// Configure the Git author name & email for this repository (or globally)
pub fn set_repo_identity(
    repo: &Repository,
    name: &str,
    email: &str,
    is_global: bool,
) -> AppResult<()> {
    let name = name.trim();
    let email = email.trim();

    if name.is_empty() {
        return Err(AppError::InvalidRepo("Tên tác giả không được để trống.".into()));
    }
    if email.is_empty() {
        return Err(AppError::InvalidRepo("Email tác giả không được để trống.".into()));
    }

    if is_global {
        let mut global_cfg = repo.config()?.open_global()?;
        global_cfg.set_str("user.name", name)?;
        global_cfg.set_str("user.email", email)?;
    } else {
        let mut cfg = repo.config()?;
        cfg.set_str("user.name", name)?;
        cfg.set_str("user.email", email)?;
    }

    Ok(())
}
